src/NHibernate/Cache/CacheKey.cs

```diff
@@ -50,7 +56,22 @@ public override bool Equals(object obj)
 
 		public override int GetHashCode()
 		{
-			return hashCode;
+			// If the object is put in a set or dictionary during deserialization, the hashcode will not yet be
+			// computed. Compute the hashcode on the fly. So long as this happens only during deserialization, there
+			// will be no thread safety issues. For the hashcode to be always defined after deserialization, the
+			// deserialization callback is used.
+			return _hashCode ?? GenerateHashCode();
```

Comment:
I could have written it `_hashCode ?? (_hashCode = GenerateHashCode()).Value`, for lazily initializing it. But that is slightly more complicated code for a corner case, which should rarely happen: deserialization in which the object is put in a set or dictionary as part of the deserialization process. The constructor and the deserialization callback are the normal paths of initialization of this member.
They are mandatory, because solely relying on a lazy initialization could expose the code of `GenerateHashCode` to multi-threaded usage. It should work in most cases, but better reduce as much as possible the code having to support multi-threading.

Reply:
I have some comments at QueryKey and FilterKey, but i test it and it works.
The cache problem will be fixed with this pr.